Add Jetpack performance testing CI infrastructure

LiamSarsfield opened the pull request:
Addresses [HOG-438: Create Jetpack Performance Tooling for LCP](https://linear.app/a8c/issue/HOG-438/create-jetpack-performance-tooling-for-lcp)

## Proposed changes:

* Add performance testing infrastructure under `tools/performance/` to measure wp-admin dashboard LCP (Largest
Contentful Paint) with Jetpack connected
* Uses Docker for isolated WordPress environment with simulated WordPress.com connection (fake tokens + mocked API
with 200ms latency)
* Includes CPU throttling calibration for consistent results across different machines
* Posts metrics to CodeVitals for tracking over time

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with
  a script to run)?

## Jetpack product discussion

pc9hqz-3Rb-p2

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:

Prerequisites: Docker running, Node 18+

```bash
cd tools/performance
pnpm install
pnpm exec playwright install chromium
pnpm calibrate
pnpm test -- --skip-codevitals
```

The test suite will automatically clone the pre-built plugin from [jetpack-production](https://github.com/Automattic/jetpack-production) on first run.

Expected output: LCP measurement for wp-admin dashboard with Jetpack connected (simulated)

LiamSarsfield committed:
Add Jetpack performance testing CI infrastructure

Introduces automated LCP (Largest Contentful Paint) measurement for the
wp-admin dashboard with simulated Jetpack WordPress.com connection.

Key components:
- Docker environment with WordPress + simulated Jetpack connection
- CPU throttling calibration for consistent results across CI agents
- Playwright-based LCP measurement
- CodeVitals integration for metric tracking

Metric posted: wp-admin-dashboard-connection-sim-largestContentfulPaint

LiamSarsfield committed:
Trim verbose JSDoc and remove dead code from performance scripts

- Remove outdated 4-scenario comment (only 1 scenario exists)
- Remove dead baseline comparison code that never executed
- Trim JSDoc to single-line descriptions across all scripts
- Update eslint config to allow minimal JSDoc

LiamSarsfield committed:
Fix eslint import/no-unresolved for standalone dependencies

LiamSarsfield committed:
Fix race condition with sequential container startup

Changed the Docker startup sequence to prevent race conditions where
WordPress containers interfere with WP-CLI's database operations:

1. Start only the db container
2. Wait for MySQL to be ready
3. Run WP-CLI setup (WordPress container NOT running)
4. Start WordPress containers

This ensures WP-CLI has exclusive database access during setup,
eliminating "table doesn't exist" errors caused by concurrent access.

Changes:
- run-performance-tests.js: Sequential container startup
- setup-wordpress.sh: Simplified (removed HTTP wait logic)
- docker-compose.yml: Removed wpcli depends_on wordpress

LiamSarsfield committed:
Fix ESLint config: remove import/no-unresolved rule

The import plugin is not available in this context, and the base config
handles import resolution. Also ensure jsdoc rules are disabled for
these utility scripts.

LiamSarsfield committed:
Disable import/no-unresolved rule for performance scripts

Dependencies (playwright, dotenv) are in tools/performance/package.json,
not the monorepo root, so the import resolver can't find them.

LiamSarsfield committed:
Add detailed calibration verification logging

- Log calibration file path and existence
- Show throttle rate, target score, calibration time, and samples
- Confirm throttling is applied via CDP on first iteration

LiamSarsfield committed:
Add tools/performance

LiamSarsfield committed:
Update pnpm-lock.yaml to include dotenv and playwright dependencies for tools/performance

LiamSarsfield committed:
Update docker-compose.yml to change the dependency condition for the wordpress-jetpack-connected service, ensuring it starts only after the service is initiated, which helps prevent race conditions during container startup.

LiamSarsfield committed:
Enhance performance testing documentation and scripts

- Updated README.md to include new environment variables: CODEVITALS_URL, GIT_BRANCH, WP_ADMIN_USER, and WP_ADMIN_PASS.
- Modified post-to-codevitals.js to streamline metric extraction by removing unused baseMetrics.
- Improved run-performance-tests.js to prioritize GIT_COMMIT environment variable for git hash retrieval, ensuring accurate tracking during CI runs.

LiamSarsfield committed:
Add baseMetrics placeholder in post-to-codevitals.js

- Introduced an empty baseMetrics object in the payload to clarify that baseline normalization is not utilized in the performance metrics submission.

LiamSarsfield committed:
Refactor browser launch configuration in calibrate-throttling.js

- Simplified browser launch to always use headless mode for consistency in performance calibration.
- Removed conditional logic for headful mode, ensuring a streamlined execution in both local and CI environments.

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/perf-testing-ci-mvp

anomiex commented on tools/performance/README.md:
```diff
@@ -0,0 +1,51 @@
+
+## CI Usage
+
+The test suite is designed to run in TeamCity. See build configuration for setup.
```
I wonder if it'd make more sense to run it in Actions rather than TeamCity, on each commit to trunk instead of backfilling weekly.

If you're running it in TeamCity, are you looking at the monorepo, or at https://github.com/Automattic/jetpack-production which has the already-built plugin?

kraftbj replied:
One of the exploration intents is to see about reusing work. With our GHE instance, TeamCity is the runner we have to use, so seeing the pros/cons of TC here.

LiamSarsfield replied:
> I wonder if it'd make more sense to run it in Actions rather than TeamCity, on each commit to trunk instead of backfilling weekly.

GitHub Actions runners can have unpredictable performance variability due to shared infrastructure. For detecting small regressions (5-10%), this noise can drown out minor regressions over time. The TeamCity build runs on a dedicated machine with no other agents interfering.

> If you're running it in TeamCity, are you looking at the monorepo, or at [Automattic/jetpack-production](https://github.com/Automattic/jetpack-production?rgh-link-date=2025-12-15T16%3A03%3A12Z) which has the already-built plugin?

The monorep is used instead of the already-built plugin mainly so we can leverage commit level tracking, so we can have the ability to bisect when issues appear (where needed). It may not be needed for now though, but you can see [within CodeVitals ](https://www.codevitals.run/project/jetpack/wp-admin-dashboard-connection-sim-largestContentfulPaint)that a point can be clicked and we can see the commit details when clicking on the commit SHA.

<img width="354" height="53" alt="Screenshot 2025-12-18 at 14 37 07" src="https://github.com/user-attachments/assets/8fb01d91-8260-45aa-ab70-54d87191edf3" />

anomiex replied:
Note that every commit that changes anything in Jetpack gets mirrored to [Automattic/jetpack-production](https://github.com/Automattic/jetpack-production), absent rare cases where something goes wrong with the mirroring. Each commit there also includes a footer like `Upstream-Ref: Automattic/jetpack@d5b54134bb471f3a54d04b12e128e6e0e2d77bde` to make it easy to find the corresponding monorepo commit.

It's up to you, but it may save having to build each monorepo commit before you can test it.

It'll also either save you testing commits that don't affect Jetpack-the-plugin at all or save you having to maintain code to try to identify which ones do. I see your P2 post mentions

> that touched PHP files in Jetpack or packages

That would unnecessarily test changes to packages like `packages/jetpack-mu-wpcom`, unless you have a list of packages to ignore. It could also miss JS-only changes that might still affect the LCP timing.

LiamSarsfield replied:
> Note that every commit that changes anything in Jetpack gets mirrored to [Automattic/jetpack-production](https://github.com/Automattic/jetpack-production?rgh-link-date=2025-12-18T14%3A57%3A35Z), absent rare cases where something goes wrong with the mirroring. Each commit there also includes a footer like Upstream-Ref: Automattic/jetpack@d5b54134bb471f3a54d04b12e128e6e0e2d77bde to make it easy to find the corresponding monorepo commit.

Actually yes! I should have checked this before hand, I didn't realise `Automattic/jetpack-production` has bissectable commits like this. It'll also help regarding the issue you mentioned for testing commits that don't affect Jetpack-the-plugin also. Thanks, I'll update the TeamCity build to use that instead.

LiamSarsfield replied:
Updated to use jetpack-production, the implementation now:

  1. Clones jetpack-production instead of building from the monorepo
  2. Parses Upstream-Ref from each mirror commit to track the original monorepo SHA in CodeVitals
  3. Removes file filtering from the TC scheduler

  The monorepo VCS root is still used for the tools/performance/ scripts, but the plugin itself comes directly from the pre-built mirror.

anomiex commented on tools/performance/eslint.config.mjs:
```diff
@@ -0,0 +1,25 @@
+			'jsdoc/require-param-type': 'off',
+			// Dependencies are in tools/performance/package.json, not monorepo root
+			// so the import resolver can't find them. Disable this rule.
+			'import/no-unresolved': 'off',
```
That doesn't matter, eslint will look in `tools/performance/node_modules` for files under here. 

Also, when I remove this and run eslint, it doesn't seem to complain about anything.

LiamSarsfield replied:
Nice catch! Removed the import/no-unresolved rule and its comment. Also simplified the config to match [tools/cli/eslint.config.mjs](https://github.com/Automattic/jetpack/blob/trunk/tools/cli/eslint.config.mjs#L1)

anomiex reviewed (commented):
Seems ok from a monorepo perspective. I didn't look too closely at the code. 

If you're wanting a review of the faked-connection stuff in the mu-plugin, @Automattic/jetpack-vulcan would be the team to ask.

kraftbj reviewed (dismissed):
I'm not sure where it came from; I was given a noticed that MacOS's `rsync` doesn't handle symlinks well and stating I should `brew install rsync`. It let me proceed, but it failed.

Installing `rsync` via `brew install rsync` and trying again worked. I don't think it's a blocker but wanted to note it.

LiamSarsfield committed:
Refactor ESLint configuration for performance testing CLI scripts

- Consolidated rules into a single object for improved readability and maintainability.
- Removed unnecessary comments and streamlined the configuration structure.

LiamSarsfield committed:
Update performance testing setup and documentation

- Added 'plugin/' directory to .gitignore to exclude cloned plugin files.
- Updated README.md with detailed setup instructions and clarified usage of the pre-built Jetpack plugin.
- Modified docker-compose.yml to mount the plugin from the new directory structure.
- Refactored run-performance-tests.js to clone the Jetpack plugin from the production mirror instead of using rsync, ensuring a more straightforward setup process.

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/perf-testing-ci-mvp

LiamSarsfield committed:
Enhance Jetpack_WPCom_Connection_Simulator to log unhandled endpoints

- Introduced a mechanism to log unhandled endpoints, aiding in the identification of missing mock responses.
- Added a flag to track if a specific endpoint handler was matched, improving response handling clarity.
- Updated comments for legacy endpoints to indicate early returns without logging.

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/perf-testing-ci-mvp

LiamSarsfield committed:
Merge branch 'trunk' into add/perf-testing-ci-mvp
